src/controllers/group.js

```diff
@@ -353,6 +354,51 @@ const Group = {
     return res.status(201).send({ images });
   },
 
+  deleteImages: async (req, res, next) => {
+    // array of image ids
+    let { images } = req.body;
+    const { id } = req.params;
+    let group;
+
+    try {
+      group = await GroupModel.findById(id).exec();
+    } catch (err) {
+      return next(new APIError(
+        undefined,
+        undefined,
+        undefined,
+        err,
+      ));
+    }
+
+    if (!group) {
+      return next(new APIError(
+        'Could not find Group',
+        `Group with id ${id} could not be found.`,
+        404,
+        `/groups/${id}`,
+      ));
+    }
+
+    try {
+      images = await ImageModel.find({ $and: [{ _id: { $in: images } }, { groupID: id }] }).exec();
+
+      await Promise.all(images.map(async (image) => {
+        S3.deleteObject(image.key);
```

Comment:
I think this needs to be prefix with `await`

Reply:
It doesn't since it will return a promise, which will be resolved by `Promise.all`

Reply:
it would only resolve the return value not the promises within the block. Technically in your case since you're not using `await` you map function doesn't need to be async.

Reply:
The comment on the async function I agree with, but `map` will return an array of promises, which `Promise.all` resolves, right? If I put await inside the block, we won't utilize the async nature and end up doing the calls practically sequentially.

Reply:
No, when using `await` each **block** will run in parallel but the **statements** within the block will run sequentially. As it stands right now you're calling `image.deleteOne();` before knowing if ` S3.deleteObject(image.key);` properly finished.

Reply:
Why is that a problem though? Won't both promises be pushed to the resultant array from map?

Reply:
No, because you don't return anything from the map. The Promise.all() isn't doing anything because nothing is being `await`'ed

Reply:
Do both statements need to be awaited?

Reply:
You do either:
```js
.map(item => {
  await myAsyncFunc1();
  return myAsyncFunc2();
})
```

Or
```js
.map(item => {
  await myAsyncFunc1();
  await myAsyncFunc2();
})
```
Both operate the same way.

Reply:
Ah I see. So each iteration of map runs in parallel, but for each iteration, the statements run sequentially? Also, could another solution be to push the promises into an array and then return that array, which is then pushed into the array returned by map? Could this be more efficient since the sequential statements then also can run without waiting for async statements?

Reply:
> So each iteration of map runs in parallel, but for each iteration, the statements run sequentially?

If they're awaited, yeah. If not statement B might execute before statement A finished.

> Also, could another solution be to push the promises into an array and then return that array, which is then pushed into the array returned by map?

This could work in resolving the promises properly but there's still no guarantee that statement A will finish before statement B. Since they're both in an array and treated as equal in terms of precedence.

> Could this be more efficient since the sequential statements then also can run without waiting for async statements?

The only way to be truly sequential is to `await` on async statements. You can't be both sequential and not "wait" for another previous async statement to finish.

Reply:
I still don't see why we care what order they are finished in.

Reply:
If you don't care the order they're finished and/or the result of the mapping, then you don't need to use Promise.all() or and async block.

Reply:
But then couldn't that lead the to function returning before the operations are finished?

Reply:
They're just functions without callbacks at that point. They start executing their async functions in the background *even after the function returns*.

Reply:
Ah alright then, that's what I was going for. I'll just remove the promise.all.

Reply:
I'm pretty sure my statements were correct, but I would just test it out to verify.